src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
+
+            var updateSelectionHint = GetUpdateSelectionHint(bestItem, snapshot, caretPosition, triggerReason, filterText, selectedItemIndex, mustSetSelection);
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
+
+            return new EditorCompletion.FilteredCompletionModel(highlightedList, selectedItemIndex, updatedFilters, updateSelectionHint, centerSelection: true, uniqueItem);
+        }
+
+        private static RoslynCompletionItem GetOrCreateRoslynItem(EditorCompletion.CompletionItem item)
+        {
+            if (item.Properties.TryGetProperty<RoslynCompletionItem>(CompletionItemSource.RoslynItem, out var roslynItem))
+            {
+                return roslynItem;
+            }
+
+            return RoslynCompletionItem.Create(item.DisplayText, item.FilterText, item.SortText);
+        }
+
+        private EditorCompletion.UpdateSelectionHint GetUpdateSelectionHint(
+            RoslynCompletionItem bestItem, ITextSnapshot snapshot, int? caretPosition, 
+            EditorCompletion.CompletionTriggerReason triggerReason, string filterText, int selectedItemIndex, bool mustSetSelection)
+        {
+            if (bestItem == null || mustSetSelection)
+            {
+                return EditorCompletion.UpdateSelectionHint.SoftSelected;
+            }
+
+            // We don't have a builder and we have a best match.  Normally this will be hard
+            // selected, except for a few cases.  Specifically, if no filter text has been
+            // provided, and this is not a preselect match then we will soft select it.  This
+            // happens when the completion list comes up implicitly and there is something in
+            // the MRU list.  In this case we do want to select it, but not with a hard
+            // selection.  Otherwise you can end up with the following problem:
+            //
+            //  dim i as integer =<space>
+            //
+            // Completion will comes up after = with 'integer' selected (Because of MRU).  We do
+            // not want 'space' to commit this.
+
+            if (ShouldSoftSelectItem(bestItem, filterText, triggerReason))
+            {
+                return EditorCompletion.UpdateSelectionHint.SoftSelected;
+            }
+
+            // If the user moved the caret left after they started typing, the 'best' match may not match at all
+            // against the full text span that this item would be replacing.
+            if (!MatchesFilterText(bestItem.FilterText, bestItem.DisplayText, bestItem.Rules.MatchPriority, filterText, triggerReason))
+            {
+                return EditorCompletion.UpdateSelectionHint.SoftSelected;
+            }
+
+            if (caretPosition == bestItem.Span.Start && bestItem.Span.Length > 0)
```

Comment:
more comments were removed.  why!?

These comments are seriously important and really ensure that people reading and maintaining the code know what's happening.  I can't look at this line and have any idea why it's doing what it's doing.  Previously, i would at least have *some* idea why we had this code and what cases it was trying to support.